modmail/extensions/threads.py

```diff
@@ -0,0 +1,459 @@
+from modmail.utils.threads.errors import ThreadAlreadyExistsError, ThreadNotFoundError
+from modmail.utils.users import check_can_dm_user
+
+
+if TYPE_CHECKING:
+    from modmail.bot import ModmailBot
+    from modmail.log import ModmailLogger
+
+EXT_METADATA = ExtMetadata()
+
+
+BASE_JUMP_URL = "https://discord.com/channels"
+USER_NOT_ABLE_TO_BE_DMED_MESSAGE = (
+    "**{0}** is not able to be dmed! This is because they have either blocked the bot, "
+    "or they are only accepting direct messages from friends.\n"
+    "It is also possible that they are not in the same server as the bot. "
+)
+ON_SUCCESS_EMOJI = "\u2705"  # ✅
+
+# This will be part of configuration later, so its stored in globals for now
+ENABLE_DM_TO_GUILD_TYPING = False  # Library bug prevents this form working right now
+ENABLE_GUILD_TO_DM_TYPING = False
+
+USE_AUDIT_LOGS = True
+logger: "ModmailLogger" = logging.getLogger(__name__)
+
+
+class TicketsCog(ModmailCog, name="Threads"):
+    """A cog for relaying direct messages."""
+
+    def __init__(self, bot: "ModmailBot"):
+        self.bot = bot
+        # user id, Ticket
+        self.relay_channel: Union[
+            discord.TextChannel, discord.PartialMessageable
+        ] = self.bot.get_partial_messageable(self.bot.config.thread.relay_channel_id)
+        self.embed_creator = ThreadEmbed()
+        self.thread_create_delete_lock = asyncio.Lock()
+
+        self.use_audit_logs: bool = USE_AUDIT_LOGS
+        self.fetch_necessary_values.start()
+
+    async def init_relay_channel(self) -> None:
+        """Fetch the relay channel."""
+        self.relay_channel = await self.bot.fetch_channel(self.bot.config.thread.relay_channel_id)
+
+    @tasks.loop(count=1)
+    async def fetch_necessary_values(self) -> None:
+        """Fetch the audit log permission."""
+        self.relay_channel: discord.TextChannel = await self.bot.fetch_channel(self.relay_channel.id)
+        self.relay_channel.guild = await self.bot.fetch_guild(self.relay_channel.guild.id)
+        me = await self.relay_channel.guild.fetch_member(self.bot.user.id)
+        self.use_audit_logs = USE_AUDIT_LOGS & me.guild_permissions.view_audit_log
+        logger.debug("Fetched relay channel and use_audit_log perms")
+
+    def cog_unload(self) -> None:
+        """Cancel any tasks that may be running on unload."""
+        self.fetch_necessary_values.cancel()
+
+    def get_ticket(self, id: int, /) -> Ticket:
+        """Fetch a ticket from the tickets dict."""
+        try:
+            ticket = self.bot.tickets[id]
+        except KeyError:
+            raise ThreadNotFoundError(f"Could not find thread from id {id}.")
+        else:
+            return ticket
+
+    # the reason we're checking for a user here rather than a member is because of future support for
+    # a designated server to handle threads and a server where the community resides,
+    # so its possible that the user isn't in the server where this command is run.
+    @commands.command()
```

Comment:
Could we move this below `_relay_message`, so it is in the "commands" region?

Reply:
I would rather keep the commands decorator not split from the function, since its part of the function itself.